sdk/eventhub/event-hubs/src/partitionLoadBalancer.ts

```diff
@@ -117,58 +117,61 @@ export class FairPartitionLoadBalancer implements PartitionLoadBalancer {
-   * and no other event processor owns lesser number of partitions than this event processor.
-   */
+  
   private _shouldOwnMorePartitions(
-    ourOwnerId: string,
-    minPartitionsPerEventProcessor: number,
-    partitionIds: string[],
-    ownerPartitionMap: Map<string, PartitionOwnership[]>
+    minRequired: number,
+    numEventProcessorsWithAdditionalPartition: number,
+    numPartitionsOwnedByUs: number,
+    processorCounts: ProcessorCounts
   ): boolean {
-    const numberOfPartitionsOwned = ownerPartitionMap.get(ourOwnerId)!.length;
+    let maxPartitionsPerProcessor = minRequired;
 
-    let sumOfPartitionsOwnedByAnyProcessor = 0;
-    for (const ownershipList of ownerPartitionMap.values()) {
-      sumOfPartitionsOwnedByAnyProcessor =
-        sumOfPartitionsOwnedByAnyProcessor + ownershipList.length;
+    if (
+      numEventProcessorsWithAdditionalPartition > 0 &&
+      processorCounts.haveExtraPartition < numEventProcessorsWithAdditionalPartition
+    ) {
+      maxPartitionsPerProcessor = minRequired + 1;
     }
+
+    return numPartitionsOwnedByUs >= maxPartitionsPerProcessor;
+  }
+
+  private _isLoadBalanced(
+    requiredCountWithAdditionalPartition: number,
+    totalPartitions: number,
+    processorCounts: ProcessorCounts
+  ): boolean {
     return (
-      numberOfPartitionsOwned < minPartitionsPerEventProcessor ||
-      (sumOfPartitionsOwnedByAnyProcessor < partitionIds.length &&
-        numberOfPartitionsOwned < minPartitionsPerEventProcessor + 1)
+      processorCounts.haveExtraPartition === requiredCountWithAdditionalPartition &&
+      processorCounts.haveRequiredPartitions + processorCounts.haveExtraPartition === totalPartitions
     );
   }
 
-  /*
-   * When the load is balanced, all active event processors own at least minPartitionsPerEventProcessor
-   * and only numberOfEventProcessorsWithAdditionalPartition event processors will own 1 additional
-   * partition.
-   */
-  private _isLoadBalanced(
-    minPartitionsPerEventProcessor: number,
-    numberOfEventProcessorsWithAdditionalPartition: number,
+  private _getProcessorCounts(
+    numRequired: number,
     ownerPartitionMap: Map<string, PartitionOwnership[]>
-  ): boolean {
-    let count = 0;
+  ): ProcessorCounts {
+    const counts: ProcessorCounts = {
+      haveRequiredPartitions: 0,
+      haveExtraPartition: 0
+    };
+
     for (const ownershipList of ownerPartitionMap.values()) {
       const numberOfPartitions = ownershipList.length;
-      if (
-        numberOfPartitions < minPartitionsPerEventProcessor ||
-        numberOfPartitions > minPartitionsPerEventProcessor + 1
-      ) {
-        return false;
+
+      if (numberOfPartitions === numRequired) {
+        counts.haveRequiredPartitions++;
       }
-      if (numberOfPartitions === minPartitionsPerEventProcessor + 1) {
-        count++;
+
+      if (numberOfPartitions === numRequired + 1) {
```

Comment:
I'm not exactly sure what `ProcessorCounts` is meant to give you (doc comment?), but it might be better to make this:
```suggestion
      if (numberOfPartitions > numRequired) {
```
Reason being, if one or more processors were running and have claimed all the partitions, and a new one is trying to join in, the already running processors may have more than `numRequired + 1`, and you'll end up with `ProcessorCounts` where both values are 0.

Reply:
It looks like this might work correctly for the `isLoadBalanced` method, but not `shouldOwnMorePartitions`.